### ISSUE TITLE
Cycle detected between modules with ctors/dtors

### ISSUE THREAD

**Issue by Maaaks** [PhilippeSigaud/Pegged#159]
I'm getting this message when running my project using Pegged (`ltext` is the project's name, `ltext.parser` is the main class containing Pegged mixin, and `ltext.params` contains some aux classes):

> object.Exception@src/rt/minfo.d(162): Aborting: Cycle detected between modules with ctors/dtors:
> ltext.params -> ltext.parser -> ltext.params

I tried to rearrange code a little — just moved `mixin(grammar(import(...)))` into another place of the same file — and got this:

> object.Exception@src/rt/minfo.d(162): Aborting: Cycle detected between modules with ctors/dtors:
> pegged.parser -> pegged.dynamic.grammar -> pegged.parser

What could it possibly mean? What should I check to make my project work?

**Comment by PhilippeSigaud** [comment]
I never got this error. What D compiler are you using?

**Comment by Maaaks** [comment]
Compiler is DMD 2.067.1 on Ubuntu 15.04.
Besides Pegged, the project also uses Temple and Vibe.dn but I'm not sure whether it's relevant...

I can also say that I am compiling with `--build-mode=singleFile` when I get this. Can it somehow cause the message?..

**Comment by PhilippeSigaud** [comment]
On Mon, Jun 29, 2015 at 6:09 PM, Maaaks <notifications@github.com> wrote:

> Compiler is DMD 2.067.1 on Ubuntu 15.04.
>

OK. I have the same compiler and the same OS, so we're good here.



> Besides Pegged, the project also uses Temple and Vibe.dn but I'm not sure
> whether it's relevant...
>



No, there should be no link between these libraries.



> I can also say that I am compiling with --build-mode=singleFile when I
> get this. Can it somehow cause the message?.
>

I don't know this flag and DMD does not seem to know it either. What
command are you using to compile?

**Comment by Maaaks** [comment]
Oops, sorry, I meant the dub's parameter `--build-mode=singleFile`.

**Comment by PhilippeSigaud** [comment]
OK. So I guess that's why I never saw this error, I'm just using rdmd to
test pegged.

Does your project build without this flag?

**Comment by Maaaks** [comment]
Yes, but it takes almost all memory on my computer to build it: I have too many Temple templates in my project. That's why I decided to place all them into separate files and make the compiler never work with more than one template at a time. This increased compilation time a lot, but at least now I don't have to quit everything other when I'm building the project. %)So, I would like to build the project with this flag. Is it possible?                                                                                                                                                                                                                                                                                                                                                     From: Philippe SigaudSent: понедельник, 29 июня 2015 г., 22:53To: PhilippeSigaud/PeggedReply To: PhilippeSigaud/PeggedCc: MaaaksSubject: Re: [Pegged] Cycle detected between modules with ctors/dtors (#159)OK. So I guess that's why I never saw this error, I'm just using rdmd to
test pegged.

Does your project build without this flag?


—Reply to this email directly or view it on GitHub.

**Comment by Maaaks** [comment]
**I found my problem!** It is the stupidest problem ever, and it seems to have nothing to do with Pegged.

For some reason I just don't like to have separate `sources` or `src` directory inside project's root directory. Though, some of subdirectories contained submodules of my project, like this:
```json
{
	"name": "myproject",
	"sourcePaths": ["."],
	"subPackages": [
		"./submodule1",
		"./submodule2"
	],
	"dependencies": {
		"myproject:submodule1": ">=0.1.0",
		"myproject:submodule2": ">=0.1.0",
	},
}

```
But I haven't realized that the submodules' sources were in fact compiled twice: as a submodules' sources and as the main project's sources! In fact, I don't know why the linker haven't gone crazy then already, so I didn't notice that I do it all incorrectly.

When I started to use `--build-mode=singleFile`, I got some new errors, and by the way I noticed that I include the files of `submodule1` as
```d
import myproject.submodule1.something;
```
instead of
```d
import submodule1.something;
```

I fixed this, but I suppose the linker went crazy because of existense of identically named modules. Now, I forbade Dub to include the submodules' files into main project's compilation:
```json
	"excludedSourceFiles": [
		"submodule1/*",
		"submodule2/*"
	],
```
...and the problem seems to be solved!

Sorry that I thought Pegged was the reason. :-)

**Comment by PhilippeSigaud** [comment]
> *I found my problem!*
>

Good, that's good news.

I fixed this, but I suppose the linker went crazy because of existense of
> identically named modules. Now, I forbade Dub to include the submodules'
> files into main project's compilation:
>
>     "excludedSourceFiles": [
>         "submodule1/*",
>         "submodule2/*"
>     ],
>
> ...and the problem seems to be solved!
>

Good to know. I admit not being a pro of dub, so I couldn't help you there.


> Sorry that I thought Pegged was the reason. :-)
>

No problem. Sometime a new version of DMD finds a new bug, so that could
have been Pegged.

Happy coding!

**Comment by Maaaks** [comment]
You now, I'm an idiot or something, because my code doesn't compile again now. o_O

Still trying...

**Comment by Maaaks** [comment]
Some googling tells me that this is a problem that happens sometimes but it doesn't tell why.

For now, I don't see it again. I removed all the `*.o` and `*.a` files found, so maybe the problem was caused by incorrect linking (maybe stupid Dub didn't recompile a library and linked something old into the new binary?..)

In fact, I don't understand anything at all now, so I just can make you sure that it is not Pegged's fault, and I can say nothing more. :-/
